### PR TITLE
add genesis data to es_objects

### PR DESCRIPTION
**Pull request by oxarbitrage**
For issue https://github.com/bitshares/bitshares-core/issues/1652

Provides option `es-objects-load-genesis` enabled by default to save accounts, assets and balances from the genesis file to ES.

### CODE REVIEW

File: libraries/plugins/es_objects/es_objects.cpp [hunk 1]
```diff
@@ -76,10 +78,64 @@ class es_objects_plugin_impl
...
-bool es_objects_plugin_impl::index_database( const vector<object_id_type>& ids, std::string action)
+void es_objects_plugin_impl::genesis()
+{
+   ilog("elasticsearch OBJECTS: inserting data from genesis");
+
+   graphene::chain::database &db = _self.database();
+
+   block_number = 1;
+   auto block1 = db.fetch_block_by_number(1);
+   if(block1.valid())
```
Comment (pmconrad) [inline libraries/plugins/es_objects/es_objects.cpp]: Better always use `head_block_num` and `head_block_time`

File: libraries/plugins/es_objects/es_objects.cpp [hunk 2]
```diff
@@ -260,6 +316,7 @@ void es_objects_plugin::plugin_set_program_options(
          ("es-objects-index-prefix", boost::program_options::value<std::string>(), "Add a prefix to the index(objects-)")
          ("es-objects-keep-only-current", boost::program_options::value<bool>(), "Keep only current state of the objects(true)")
          ("es-objects-start-es-after-block", boost::program_options::value<uint32_t>(), "Start doing ES job after block(0)")
+         ("es-objects-load-genesis", boost::program_options::value<bool>(), "Load genesis data to ES(true)")
```
Comment (pmconrad) [inline libraries/plugins/es_objects/es_objects.cpp]: The documentation string is misleading. The option will index whatever is in the database on startup.
Also, `genesis()` is called after the first change, which includes changes that were applied after genesis.

**Comment by pmconrad** [comment]
How about connecting `database::applied_block` to `genesis()`? You could use block number 1 as the trigger, which is *almost* identical to genesis state. It is also triggered during replay, unlike `new`/`changed`/`removed_objects` .

**Comment by oxarbitrage** [comment]
Good idea, i think there is going to be less problems that way as we make sure genesis is only done at block 1.

Applied at https://github.com/bitshares/bitshares-core/pull/1717/commits/5c2692d637e6d2c0b0aa656cc7a80f197bf2b975

File: libraries/plugins/es_objects/es_objects.cpp [hunk 3]
```diff
@@ -78,60 +78,61 @@ class es_objects_plugin_impl
...
+      graphene::utilities::ES es;
+      es.curl = curl;
+      es.bulk_lines = bulk;
+      es.elasticsearch_url = _es_objects_elasticsearch_url;
+      es.auth = _es_objects_auth;
+      if (!graphene::utilities::SendBulk(std::move(es)))
+         FC_THROW_EXCEPTION(graphene::chain::plugin_exception, "Error inserting genesis data.");
+      else {
+         bulk.clear();
+         return true;
```
Comment (pmconrad) [inline libraries/plugins/es_objects/es_objects.cpp]: This return is superfluous

File: libraries/plugins/es_objects/es_objects.cpp [hunk 4]
```diff
@@ -260,12 +313,21 @@ void es_objects_plugin::plugin_set_program_options(
          ("es-objects-index-prefix", boost::program_options::value<std::string>(), "Add a prefix to the index(objects-)")
          ("es-objects-keep-only-current", boost::program_options::value<bool>(), "Keep only current state of the objects(true)")
          ("es-objects-start-es-after-block", boost::program_options::value<uint32_t>(), "Start doing ES job after block(0)")
+         ("es-objects-load-genesis", boost::program_options::value<bool>(), "Index whatever is in the database on startup."
```
Comment (pmconrad) [inline libraries/plugins/es_objects/es_objects.cpp]: Does this option make much sense? Under which circumstances would you want to turn that off?

Comment (oxarbitrage) [reply libraries/plugins/es_objects/es_objects.cpp]: Not needed, removed.

File: libraries/plugins/es_objects/es_objects.cpp [hunk 5]
```diff
@@ -266,6 +315,13 @@ void es_objects_plugin::plugin_set_program_options(
 
 void es_objects_plugin::plugin_initialize(const boost::program_options::variables_map& options)
 {
+   database().applied_block.connect([&](const signed_block &b) {
```
Comment (pmconrad) [inline libraries/plugins/es_objects/es_objects.cpp]: Please capture only what's required.

**Pull request merged**